File: 1838-frequency-of-the-most-frequent-element/1838-frequency-of-the-most-frequent-element.py
class Solution:
    def maxFrequency(self, nums: List[int], k: int) -> int:
        nums.sort()
        start = 0
        maxfreq = 0
        presum = 0
        
        for end in range(len(nums)):
            presum += nums[end]
            
            while (end-start+1)*nums[end] - presum > k:
                presum -= nums[start]
                start+=1
            
            maxfreq = max(maxfreq, end-start+1)
        
        return maxfreq
        
        

File: 1838-frequency-of-the-most-frequent-element/1838-frequency-of-the-most-frequent-element.java
class Solution {
    public int maxFrequency(int[] nums, int k) {
        int n = nums.length;
        long prefixSum = 0;
        int maxFreq = 1;
        System.out.println(n);
        Arrays.sort(nums);
        int start = 0;
        
        for(int end=0; end<n; end++){
            prefixSum += nums[end];
            long target = nums[end];
            while((long)((end-start+1)*target) - prefixSum > k){
                prefixSum -= nums[start];
                start++;
            }
            
            maxFreq = Math.max(maxFreq, end-start+1);
        }
        
        return maxFreq;
        
    }
}

/*

[1,2,4]
[1,3]
   e
 s 

k = 5


2*2 = 4-3 1 


*/